Rewrite symbol management system (#543)
ethteck/splat: CHANGELOG.md

# splat Release Notes

### 0.50.0

* BREAKING CHANGES:

>   * Declaring an overlay symbol but not specifying their `segment:` or `rom:` on the symbol_addrs.txt file is no loger accepted. A warning will be emitted and the symbol won't be used.

Reviewer: ```suggestion
  * Declaring an overlay symbol but not specifying its `segment:` or `rom:` in symbol_addrs is no loger accepted. A warning will be emitted and the symbol won't be used.
```

>   * Associating a symbol to a segment that does not cover the Vram or Rom address space for that segment is no longer accepted.

Reviewer: ```suggestion
  * Associating a symbol with a segment that is not within the Vram or Rom address space for that segment is no longer accepted.
```

>     * If you need to tell splat to use a symbol from a specific segment for a given segment, then use the new prioritized segment system.

Reviewer: ```suggestion
    * Use the new prioritized segment system to tell splat to use a symbol from a specific segment
```

>   * Declaring a symbol without an associated segment, which is not covered by any of the Vram addresses of the global segments is no longer accepted.

Reviewer: ```suggestion
  * Declaring a symbol without an associated segment and outside the Vram space of any global segments is no longer accepted.
```

>     * If you need to declare symbols that aren't part of any real segment of the rom then use the new absoulte symbols system.

Reviewer: ```suggestion
    * Use the new absoulte symbols system to declare symbols that aren't part of any real segment.
```

  * `splat.utils.symbols.all_symbols_dict` and `splat.utils.symbols.all_symbols_ranges` have been removed.
  * The algorithm for checking vram overlaps between global segments and overlays have been changed and it is more strict now.
    * This means it is more likely to produce errors when it didn't before.
    * A common source of errors is giving a vram address (usually `vram: 0xA4000040`) for the ipl3 segment. A recommened workaround is to either remove the vram for this segment or tag it with its own `exclusive_ram_id` (i.e. `exclusive_ram_id: ipl3`).

>   * It is pretty likely for many symbols that have autogenerated names to change their naming a bit.

Reviewer: ```suggestion
  * Autogenerated symbol names may have changed
```

>     * In the past splat failed to properly set some attributes required for autogenerated symbol naming, so splat symbolized them "wrong". For example it only used the vram address (like `D_{VRAM}`) when it should have been using both the vram and rom addresses (`D_{VRAM}_{ROM}`) for naming overlay symbols.

Reviewer: ```suggestion
    * Previously, splat failed to properly set some attributes required for autogenerated symbol naming and thus symbolized them incorrectly. For example, it only used the vram address (like `D_{VRAM}`) when it should have been using both the vram and rom addresses (`D_{VRAM}_{ROM}`) for naming overlay symbols.
```

    * Now those attributes are correctly set, meaning those symbols are correctly named now.

>     * This is breaking if you were referencing those wrongly named symbols in C code, because you'll get linker errors for not finding the old named symbols.

Reviewer: I feel like all these bullets should be moved to a caution section of the changelog and shouldn't be separate - they're all about the same idea. It can just be a new caution paragraph or something

  * Versions between 0.41.0 and 0.50.0 have been skipped as a way to note this can be a pretty big breaking change.

> * Rewrite symbol management system.

Reviewer: We're switching verb tenses here - "stuff has been done" vs "do stuff". Personally, for splat, I'd prefer the past tense for the changelog. I can do an edit pass on this if you don't mind.

Author: I would appreciate a lot if you did that! Feel free to go ahead, I kinda suck writing up stuff :sweat_smile:
